tools/syz-declextract/syz-declextract.cpp

```diff
@@ -1,279 +1,103 @@
-// The easiest way to build it is to build it as part of clang. Add the following lines to CMakeLists.txt:
-//   +add_clang_executable(syz-declextract syz-declextract/syz-declextract.cpp)
-//   +target_link_libraries(syz-declextract clangTooling)
-// It was used to extract windows descriptions:
-//   syz-declextract -extra-arg="--driver-mode=cl" -extra-arg="-I/path/to/windows/headers" Windows.h
-
-#include "clang/AST/AST.h"
-#include "clang/AST/ASTConsumer.h"
-#include "clang/AST/ASTContext.h"
-#include "clang/AST/RecursiveASTVisitor.h"
-#include "clang/Driver/Options.h"
-#include "clang/Frontend/ASTConsumers.h"
-#include "clang/Frontend/CompilerInstance.h"
-#include "clang/Frontend/FrontendActions.h"
-#include "clang/Rewrite/Core/Rewriter.h"
+#include "clang/AST/APValue.h"
+#include "clang/AST/Decl.h"
+#include "clang/AST/Expr.h"
+#include "clang/AST/Stmt.h"
+#include "clang/AST/Type.h"
+#include "clang/ASTMatchers/ASTMatchFinder.h"
+#include "clang/ASTMatchers/ASTMatchers.h"
+#include "clang/ASTMatchers/ASTMatchersInternal.h"
+#include "clang/Basic/LLVM.h"
+#include "clang/Sema/Ownership.h"
 #include "clang/Tooling/CommonOptionsParser.h"
 #include "clang/Tooling/Tooling.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/raw_ostream.h"
+#include <stdio.h>
+#include <utility>
 
 using namespace clang;
-using namespace clang::tooling;
-
-std::string convertType(ASTContext &C, QualType T) {
-  auto name = T.getAsString();
-  if (name == "HANDLE")
-    return name;
-  if (T->isIntegralOrEnumerationType()) {
-    int size = C.getTypeSize(T);
-    char buf[10];
-    sprintf(buf, "int%d", size);
-    return buf;
-  }
-  if (T->isVoidPointerType()) {
-    return "ptr[inout, array[int8]]";
-  }
-  if (T->isPointerType()) {
-    auto inner = convertType(C, T->getPointeeType());
-    if (inner == "")
-      return "ptr[inout, array[int8]]";
-    char buf[1024];
-    sprintf(buf, "ptr[inout, %s]", inner.c_str());
-    return buf;
-  }
-  return "intptr";
-}
+using namespace clang::ast_matchers;
 
-class DeclExtractCallVisitor : public RecursiveASTVisitor<DeclExtractCallVisitor> {
- public:
-  explicit DeclExtractCallVisitor(ASTContext *Context)
-      : Context(*Context) {}
-
-  bool VisitFunctionDecl(const FunctionDecl *D) {
-    if (D->doesThisDeclarationHaveABody())
-      return true;
-    // TODO(dvyukov): need to select only stdcall (WINAPI) functions.
-    // But the following 2 approaches do not work.
-    if (false) {
-      if (auto *FPT = D->getType()->getAs<FunctionProtoType>()) {
-        if (FPT->getExtInfo().getCC() != CC_X86StdCall)
-          return true;
-      }
-    }
-    if (false) {
-      if (!D->hasAttr<StdCallAttr>())
-        return true;
+class Printer : public MatchFinder::MatchCallback {
+public:
+  virtual void run(const MatchFinder::MatchResult &Result) override {
+    const VarDecl *varDecl = Result.Nodes.getNodeAs<VarDecl>("Struct");
```

Comment:
LLVM codebase makes extensive use of `auto`, I think you can do that here, too.

Reply:
Done